row.go

```diff
@@ -782,3 +784,46 @@ func reconstructFuncOfLeaf(columnIndex int16, node Node) (int16, reconstructFunc
 		return row[1:], err
 	}
 }
+
+// rowAllocator is a memory allocator used to make copy of rows referencing
+// memory buffers that parquet-go does not have ownership of.
+//
+// This type is used in the implementation of a various readers and writers that
+// need to capture rows passed to the ReadRows/WriteRows methods.
+//
+// See: RowBuffer, DedupeRowReader, DedupeRowWriter
+type rowAllocator struct{ buffer []byte }
+
+func (a *rowAllocator) reset() {
+	a.buffer = a.buffer[:0]
+}
+
+func (a *rowAllocator) capture(row Row) {
+	for i, v := range row {
+		switch kind := v.Kind(); kind {
+		case ByteArray, FixedLenByteArray:
+			row[i].ptr = unsafecast.AddressOfBytes(a.copyBytes(v.byteArray()))
+		}
+	}
+}
+
+func (a *rowAllocator) copyBytes(v []byte) []byte {
+	if free := cap(a.buffer) - len(a.buffer); free < len(v) {
+		newCap := 2 * cap(a.buffer)
+		if newCap == 0 {
+			newCap = defaultReadBufferSize
+		}
+		for newCap < len(v) {
+			newCap *= 2
+		}
+		a.buffer = make([]byte, 0, newCap)
```

Comment:
seeing lines like this make me consider changing the `bufferPool` to be a `[]byte` pool instead of a `*buffer` pool. it would allow us to recapture slices like this. wdyt?

Reply:
One issue with pooling []byte is the API of sync.Pool is based on interface{}. The compiler can optimize packing pointer values into an interface without extra heap allocation but a slice is too large for that and packing it into an interface requires making an intermediary heap allocation to hold the slice value. In an effort to reduce memory pressure we end up causing heap allocations of more small objects, depending on the access pattern this can result in just as much GC work since it tends to scale based on number of objects on the heap rather than their size. 

I want to be cautious not to introduce memory pools before having a measure of their effectiveness, they always come with a non-zero maintenance cost, manual memory management is prone to more bugs than letting the GC help us out. 

Note that these allocations are expected to be amortized across large chunks of work, applications can process millions of rows and only incur a few reallocations of these buffers. Giving the GC a bit of work is OK as long as the program scales well, it buys us safety and a lower long term maintenance cost. If we have a strong signal that the pool will be more effective we can revisit then.

Reply:
Thank you for the detailed explanation and agree that taking on pooling should come with clear evidence of a win. Appreciate the heads up on the `interface{}` heap allocations. I did not know that was optimized away for a pointer, but not a slice.